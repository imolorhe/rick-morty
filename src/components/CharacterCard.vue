<template>
  <div
    class="max-w-sm w-full lg:max-w-full lg:flex shadow-md hover:shadow-xl transition duration-300 ease-in transform hover:-translate-y-px rounded cursor-pointer"
    @click="onClickCard"
  >
    <div
      class="h-48 lg:h-auto lg:w-48 flex-none bg-cover rounded-t lg:rounded-t-none lg:rounded-l text-center overflow-hidden"
      :style="`background-image: url('${character.image}')`"
      :title="character.name"
    ></div>
    <div
      class="bg-white rounded-b lg:rounded-b-none lg:rounded-r p-4 flex flex-col flex-grow justify-between leading-normal text-left"
    >
      <div class="mb-8">
        <div class="text-gray-900 font-bold text-xl mb-2">
          {{ character.name }}
          <div class="text-sm text-gray-600">
            {{ character.gender }} &middot;
            <span
              tid="character-card-status"
              :class="statusClasses"
              >{{ character.status }}</span
            >
            &middot; {{ character.origin.name }}
          </div>
        </div>
        <p class="text-sm text-gray-600 flex items-center">
          <img src="../assets/planet.png" class="w-5 mr-1" />
          {{ character.location.name }}
        </p>
        <div class="mb-3"></div>
        <span
          class="inline-block bg-gray-200 rounded-full px-3 py-1 text-sm font-semibold text-gray-700 mr-2"
          >{{ character.species }}</span
        >
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import Vue from 'vue';

export default Vue.extend({
  name: 'CharacterCard',
  props: {
    character: {
      type: Object,
      required: true,
    },
  },
  computed: {
    statusClasses(): unknown {
      return {
        'ml-1': true,
        'text-green-600': this.character.status === 'Alive',
        'text-red-600': this.character.status !== 'Alive',
      };
    },
  },
  methods: {
    onClickCard() {
      this.$emit('click', this.character);
    },
  },
});
</script>
