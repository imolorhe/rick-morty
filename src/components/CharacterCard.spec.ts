import CharacterCard from './CharacterCard.vue';
import { shallowMount } from '@vue/test-utils';

describe('CharacterCard', () => {
  it('should render the character basic details', () => {
    const character = {
      name: 'Character',
      image: 'character.jpg',
      gender: 'Gender',
      status: 'Alive',
      species: 'Human',
      origin: {
        name: 'Origin',
      },
      location: {
        name: 'Location',
      },
    };
    const comp = shallowMount(CharacterCard, {
      propsData: {
        character,
      },
    });

    expect(comp.html()).toMatchSnapshot();
  });
  it('should render status as green if alive', () => {
    const character = {
      name: 'Character',
      image: 'character.jpg',
      gender: 'Gender',
      status: 'Alive',
      species: 'Human',
      origin: {
        name: 'Origin',
      },
      location: {
        name: 'Location',
      },
    };
    const comp = shallowMount(CharacterCard, {
      propsData: {
        character,
      },
    });
    expect(
      comp
        .find('[tid="character-card-status"]')
        .classes('text-green-600')
    ).toBe(true);
  });
  it('should render status as red if dead', () => {
    const character = {
      name: 'Character',
      image: 'character.jpg',
      gender: 'Gender',
      status: 'Dead',
      species: 'Human',
      origin: {
        name: 'Origin',
      },
      location: {
        name: 'Location',
      },
    };
    const comp = shallowMount(CharacterCard, {
      propsData: {
        character,
      },
    });
    expect(
      comp
        .find('[tid="character-card-status"]')
        .classes('text-red-600')
    ).toBe(true);
  });
  it('should emit click with the character when clicked', () => {
    const character = {
      name: 'Character',
      image: 'character.jpg',
      gender: 'Gender',
      status: 'Dead',
      species: 'Human',
      origin: {
        name: 'Origin',
      },
      location: {
        name: 'Location',
      },
    };
    const comp = shallowMount(CharacterCard, {
      propsData: {
        character,
      },
    });
    comp.trigger('click');
    expect(comp.emitted('click').length).toBe(1);
    expect(comp.emitted('click')[0]).toEqual([character]);
  });
});
