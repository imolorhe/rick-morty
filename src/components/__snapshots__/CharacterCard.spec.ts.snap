// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterCard should render the character basic details 1`] = `
<div class="max-w-sm w-full lg:max-w-full lg:flex shadow-md hover:shadow-xl transition duration-300 ease-in transform hover:-translate-y-px rounded cursor-pointer">
  <div title="Character" class="h-48 lg:h-auto lg:w-48 flex-none bg-cover rounded-t lg:rounded-t-none lg:rounded-l text-center overflow-hidden" style="background-image: url(character.jpg);"></div>
  <div class="bg-white rounded-b lg:rounded-b-none lg:rounded-r p-4 flex flex-col flex-grow justify-between leading-normal text-left">
    <div class="mb-8">
      <div class="text-gray-900 font-bold text-xl mb-2">
        Character
        <div class="text-sm text-gray-600">
          Gender ·
          <span tid="character-card-status" class="ml-1 text-green-600">Alive</span>
          · Origin
        </div>
      </div>
      <p class="text-sm text-gray-600 flex items-center"><img src="../assets/planet.png" class="w-5 mr-1">
        Location
      </p>
      <div class="mb-3"></div> <span class="inline-block bg-gray-200 rounded-full px-3 py-1 text-sm font-semibold text-gray-700 mr-2">Human</span>
    </div>
  </div>
</div>
`;
